### PR TITLE
[KHP3-6447]: Provide pre-auth UI form 

### PR DESCRIPTION
**Pull request by its-kios09**
## Requirements

- [ ] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [ ] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [ ] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

This user interface fetches outstanding invoice items and automatically fills out forms based on current visits. 

## Screenshots

*None.*

PreAuth

## Related Issue

*None.*

## Other

*None.*

### CODE REVIEW

**Comment by donaldkibet** [comment]
Could we have this work for HIE hidden under a feature flag see https://o3-docs.openmrs.org/docs/feature-flags

**Comment by its-kios09** [comment]
> Could we have this work for HIE hidden under a feature flag see https://o3-docs.openmrs.org/docs/feature-flags

Okay sure let me update it

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+      visitType: recentVisit?.visitType?.display || '',
+      facility: `${recentVisit?.location?.display || ''} - ${mflCodeValue || ''}`,
+    },
+  });
+
+  useEffect(() => {
+    setValue('diagnoses', confirmedDiagnoses);
+    setValue('visitType', recentVisit?.visitType?.display || '');
+    setValue('facility', `${recentVisit?.location?.display || ''} - ${mflCodeValue || ''}`);
+  }, [confirmedDiagnoses, recentVisit, mflCodeValue, setValue]);
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx:95]: Have we tried setting these values using react-hook-form `defaultValues`

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss [hunk 2]
```diff
@@ -0,0 +1,145 @@
+@use '@carbon/styles/scss/spacing';
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss:1]:
```suggestion
@use '@carbon/layout';
```

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+@use '@carbon/styles/scss/type';
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss:2]:
```suggestion
@use '@carbon/type';
```

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+@use '@carbon/layout';
+@import '~@openmrs/esm-styleguide/src/vars';
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss:4]: are we using this anywhere. if possible lets stick with carbon tokens

File: packages/esm-billing-app/src/claims/pre-auth/pre-auth-dashboard.component.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
+// src/components/PreAuthRequestDashboard.tsx
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/claims/pre-auth/pre-auth-dashboard.component.tsx:1]:
```suggestion

```

File: packages/esm-billing-app/src/claims/pre-auth/table/pre-auth-table.component.tsx [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+  const paginatedLineItems = useMemo(() => {
+    const startIndex = (currentPage - 1) * pageSize;
+    const endIndex = startIndex + pageSize;
+    return filteredLineItems.slice(startIndex, endIndex);
+  }, [filteredLineItems, currentPage, pageSize]);
+
+  const tableHeaders: Array<typeof DataTableHeader> = [
+    { header: 'No', key: 'no' },
+    { header: 'Serial No.', key: 'serialno' },
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/claims/pre-auth/table/pre-auth-table.component.tsx:73]: should we have these headers translated. Just to have consistency within the packages

File: packages/esm-billing-app/src/claims/pre-auth/table/pre-auth-table.component.tsx [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    { header: 'Bill Item', key: 'inventoryname' },
+    { header: 'Status', key: 'status' },
+    { header: 'Total amount', key: 'total' },
+    { header: 'Bill creation date', key: 'dateofbillcreation' },
+  ];
+
+  const processBillItem = (item) => (item.item || item.billableService)?.split(':')[1];
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/claims/pre-auth/table/pre-auth-table.component.tsx:80]: cc @ojwanganto @makombe we should not be doing this on the frontend. We need to have billable service name displayed correctly. Alternatively we can have it as an object
```json
{
  "billableService": {
    "uuid": "some-uuid",
    "display": "Some billable service"
  }
}
```

**Review by donaldkibet — COMMENTED** [review]
I've done a preliminary review. Overall LGTM. I'd recommend a more thorough review before final approval. Thank you for your work on this. cc @ojwanganto @makombe @Omoshlawi

File: packages/esm-billing-app/src/claims/pre-auth/table/pre-auth-table.component.tsx (hunk 6 above)
Comment (its-kios09) [reply packages/esm-billing-app/src/claims/pre-auth/table/pre-auth-table.component.tsx:73]: Sorted

File: packages/esm-billing-app/src/claims/pre-auth/pre-auth-dashboard.component.tsx (hunk 5 above)
Comment (its-kios09) [reply packages/esm-billing-app/src/claims/pre-auth/pre-auth-dashboard.component.tsx:1]: removed

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss (hunk 4 above)
Comment (its-kios09) [reply packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss:4]: sorted

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss (hunk 3 above)
Comment (its-kios09) [reply packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss:2]: sorted

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss (hunk 2 above)
Comment (its-kios09) [reply packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss:1]: sorted

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx (hunk 1 above)
Comment (its-kios09) [reply packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx:95]: Yes, the `default values`  in react-hook-form sets the initial default values for the form fields when you initialize the form with useForm. Or what do you think is the better approach I would appreciate your opinion

Comment (donaldkibet) [reply packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx]: This `useEffect` will force a re-render. Its best to initialize the form fields using react hook `useForm` this way we do not have a side effect

**Review by donaldkibet — CHANGES_REQUESTED** [review]
Since we're implementing multiple feature flags (e.g., pre-auth, claim management, benefits package) that all serve the same purpose of pausing HIE work, let's consolidate them into a single flag for easier management and smoother transitions.

**Comment by donaldkibet** [comment]
@its-kios09 one final bit, lets have the pre-auth form as workspace item. Ted should update the designs to follow conventions

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx (hunk 1 above)
Comment (its-kios09) [reply packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx]: i will do a follow up PR we sort it

**Comment by its-kios09** [comment]
> @its-kios09 one final bit, lets have the pre-auth form as workspace item. Ted should update the designs to follow conventions

@donaldkibet I was thinking the design I can do a follow up PR

**Review by donaldkibet — APPROVED** [review]

**Comment by donaldkibet** [comment]
Please resolve the conflicts

**Comment by donaldkibet** [comment]
Should this form be hidden behind a feature flag. If yes please hide it then we merge it in

**Review by donaldkibet — DISMISSED** [review]

**Comment by its-kios09** [comment]
> Should this form be hidden behind a feature flag. If yes please hide it then we merge it in

@donaldkibet  I have added the HIE flags

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+    },
+  });
+
+  const onSubmit = async (data) => {
+    setLoading(true);
+    try {
+      showSnackbar({
+        kind: 'success',
+        title: t('requestPreAuth', 'Pre Authorization'),
+        subtitle: t('sendClaim', 'Pre Authorization request sent successfully'),
```
Comment (makombe) [inline packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx:110]:
```suggestion
        subtitle: t('sendClaim', 'Pre authorization request sent successfully'),
```

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+        navigate({
+          to: window.getOpenmrsSpaBase() + `home/billing/`,
+        });
+      }, 2000);
+    } catch (err) {
+      console.error(err);
+      showSnackbar({
+        kind: 'error',
+        title: t('requestPreAuthError', 'Pre Authorization Error'),
+        subtitle: t('sendClaimError', 'Pre Authorization request failed, Please try later'),
```
Comment (makombe) [inline packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.component.tsx:125]:
```suggestion
        subtitle: t('sendClaimError', 'Pre authorization request failed, please try later'),
```

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      left: 0;
+    }
+
+    & > span {
+      @include type.type-style('body-01');
+    }
+  }
+
+  .buttonSet {
+    padding: 0rem;
```
Comment (makombe) [inline packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss:59]:
```suggestion
    padding: 0;
```

File: packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+.button {
+  height: layout.$spacing-08;
+  display: flex;
+  align-content: flex-start;
+  align-items: baseline;
+  min-width: 20%;
+}
+
+.buttonSet {
+  padding: 0rem;
```
Comment (makombe) [inline packages/esm-billing-app/src/claims/pre-auth/form/pre-auth-form.scss:93]:
```suggestion
  padding: 0;
```

File: packages/esm-billing-app/src/claims/pre-auth/pre-auth-dashboard.scss [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  justify-content: center;
+}
+.errorContainer {
+  margin: layout.$spacing-05;
+}
+
+.preAuthMainContainer {
+  display: flex;
+  margin: 0 auto;
+  padding: 0 1rem;
```
Comment (makombe) [inline packages/esm-billing-app/src/claims/pre-auth/pre-auth-dashboard.scss:21]: we can consistently use carbon layouts

File: packages/esm-billing-app/src/claims/pre-auth/table/pre-auth-table.component.tsx [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+            </Table>
+          </TableContainer>
+        )}
+      </DataTable>
+      {filteredLineItems?.length === 0 && (
+        <div className={styles.filterEmptyState}>
+          <Layer>
+            <Tile className={styles.filterEmptyStateTile}>
+              <p className={styles.filterEmptyStateContent}>
+                {t('noMatchingItemsToDisplay', 'No matching items to display')}
```
Comment (makombe) [inline packages/esm-billing-app/src/claims/pre-auth/table/pre-auth-table.component.tsx:213]: Maybe here we can leverage on the empty state from community provided out of the box.

**Review by makombe — COMMENTED** [review]
Generally looks good. Thanks @its-kios09

**Comment by its-kios09** [comment]
@makombe  Thanks for the suggestion

**Review by donaldkibet — APPROVED** [review]

**Comment by donaldkibet** [comment]
cc @its-kios09 merge conflicts

**Comment by ojwanganto** [comment]
I hope we still have this under a feature flag.

**Comment by its-kios09** [comment]
@donaldkibet I have sorted out the conflicts

**Pull request merged**